src/services/earn/earn-service.ts

```diff
@@ -0,0 +1,267 @@
+    private readonly allowanceService: IAllowanceService
+  ) {}
+
+  async buildCreatePositionTx({
+    chainId,
+    strategyId,
+    owner,
+    permissions,
+    strategyValidationData,
+    misc,
+    deposit,
+  }: CreateEarnPositionParams): Promise<BuiltTransaction> {
+    const vault = EARN_VAULT.address(chainId);
+    const companion = EARN_VAULT_COMPANION.address(chainId);
+    let depositInfo: { token: Address; amount: bigint; value: bigint };
+    if ('token' in deposit) {
+      const amount = BigInt(deposit.amount);
+      depositInfo = { token: deposit.token, amount, value: isSameAddress(deposit.token, Addresses.NATIVE_TOKEN) ? amount : 0n };
+    } else {
+      depositInfo = { token: deposit.permitData.token, amount: BigInt(deposit.permitData.amount), value: 0n };
+    }
+
+    const { needsSwap, asset } = await this.checkIfNeedsSwap({ chainId, strategyId, depositToken: depositInfo.token });
+
+    if ('token' in deposit && !needsSwap) {
+      // If don't need to use Permit2, then just call the vault
+      return {
+        to: vault,
+        data: encodeFunctionData({
+          abi: vaultAbi,
+          functionName: 'createPosition',
+          args: [
+            strategyId,
+            depositInfo.token as ViemAddress,
+            depositInfo.amount,
+            owner as ViemAddress,
+            permissions.map(({ operator, permissions }) => ({
+              operator: operator as ViemAddress,
+              permissions: permissions.map(mapPermission),
+            })),
+            strategyValidationData ?? '0x',
+            misc ?? '0x',
+          ],
+        }),
+        value: depositInfo.value,
+      };
+    }
+    // If we get to this point, then we'll use the Companion for the deposit
+    const calls: Hex[] = [];
+
+    // Handle take from caller (if necessary)
+    const recipient = needsSwap ? COMPANION_SWAPPER_CONTRACT.address(chainId) : companion;
+    if ('permitData' in deposit) {
+      calls.push(buildTakeFromCallerWithPermit(deposit.permitData, deposit.signature, recipient));
+    } else if (!isSameAddress(depositInfo.token, Addresses.NATIVE_TOKEN)) {
+      calls.push(buildTakeFromCaller(depositInfo.token, depositInfo.amount, recipient));
+    }
+
+    const depositToken = (needsSwap ? asset : depositInfo.token) as ViemAddress;
+    let maxApprove = false;
+    const promises = [];
+
+    if (needsSwap) {
+      const swapPromise = this.getSwapData({
+        request: {
+          chainId,
+          sellToken: depositInfo.token,
+          buyToken: asset,
+          order: { type: 'sell', sellAmount: depositInfo.amount },
+        },
+        leftoverRecipient: owner,
+        swapConfig: deposit?.swapConfig,
+      }).then(({ swapData }) => calls.push(swapData));
+      promises.push(swapPromise);
+    }
+    
+    if (!isSameAddress(depositToken, Addresses.NATIVE_TOKEN)) {
+      const allowancePromise = this.allowanceService.getAllowancesInChain({
```

Comment:
Maybe we could use `getAllowanceInChain` instead? It's easier to use for our scenario